pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go

```diff
@@ -551,3 +559,81 @@ func TestTenantStreamingCheckpoint(t *testing.T) {
 	<-time.NewTimer(3 * time.Second).C
 	require.Equal(t, [][]string{{"2"}}, c.getDestTenantSQL().QueryStr(t, "SELECT * FROM d.t2"))
 }
+
+func TestTenantStreamingUnavailableStreamAddress(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRaceWithIssue(t, 83867)
+
+	ctx := context.Background()
+	args := defaultTenantStreamingClustersArgs
+	args.srcNumNodes = 4
+	args.destNumNodes = 4
+	c, cleanup := createTenantStreamingClusters(ctx, t, args)
+	defer cleanup()
+
+	// Create a source table with multiple ranges spread across multiple nodes
+	numRanges := 10
+	rowsPerRange := 20
+	c.srcTenantSQL.Exec(t, fmt.Sprintf(`
+  CREATE TABLE d.scattered (key INT PRIMARY KEY);
+  INSERT INTO d.scattered (key) SELECT * FROM generate_series(1, %d);
+  ALTER TABLE d.scattered SPLIT AT (SELECT * FROM generate_series(%d, %d, %d));
+  ALTER TABLE d.scattered SCATTER;
+  `, numRanges*rowsPerRange, rowsPerRange, (numRanges-1)*rowsPerRange, rowsPerRange))
+
+	// Once srcCluster.Server(0) is shut down queries must be ran against a different server
+	alternateSrcExec := func(exec srcInitExecFunc) {
+		srcSysSQL := sqlutils.MakeSQLRunner(c.srcCluster.ServerConn(1))
+		_, srcTenantConn := serverutils.StartTenant(t, c.srcCluster.Server(1), base.TestTenantArgs{TenantID: c.args.srcTenantID})
+		srcTenantSQL := sqlutils.MakeSQLRunner(srcTenantConn)
+		exec(c.t, srcSysSQL, srcTenantSQL)
+	}
+
+	alternateCompareResult := func(query string) {
+		_, srcTenantConn := serverutils.StartTenant(t, c.srcCluster.Server(1), base.TestTenantArgs{TenantID: c.args.srcTenantID})
+		srcTenantSQL := sqlutils.MakeSQLRunner(srcTenantConn)
+		sourceData := srcTenantSQL.QueryStr(c.t, query)
+		destData := c.getDestTenantSQL().QueryStr(c.t, query)
+		require.Equal(c.t, sourceData, destData)
+	}
+
+	producerJobID, ingestionJobID := c.startStreamReplication()
+	jobutils.WaitForJobToRun(c.t, c.srcSysSQL, jobspb.JobID(producerJobID))
+	jobutils.WaitForJobToRun(c.t, c.destSysSQL, jobspb.JobID(ingestionJobID))
+
+	srcTime := c.srcCluster.Server(0).Clock().Now()
+	c.waitUntilHighWatermark(srcTime, jobspb.JobID(ingestionJobID))
+
+	c.destSysSQL.Exec(t, `PAUSE JOB $1`, ingestionJobID)
+	jobutils.WaitForJobToPause(t, c.destSysSQL, jobspb.JobID(ingestionJobID))
+
+	// We should've persisted the original topology
+	progress := jobutils.GetJobProgress(c.t, c.destSysSQL, jobspb.JobID(ingestionJobID))
+	topology := progress.GetStreamIngest().Topology
+	require.Greater(t, len(topology.PartitionInfo), 1)
+
+	c.srcCluster.Server(0).Stopper().Stop(ctx)
+
+	c.destSysSQL.Exec(t, `RESUME JOB $1`, ingestionJobID)
+	jobutils.WaitForJobToRun(t, c.destSysSQL, jobspb.JobID(ingestionJobID))
+
+	alternateSrcExec(func(t *testing.T, sysSQL *sqlutils.SQLRunner, tenantSQL *sqlutils.SQLRunner) {
```

Comment:
what you could try doing -- either through knobs or otherwise, is to inject a test implementation of client (wrapping real client) which augments Dial method -- just recording which nodes were dialed into and how many times; you could then assert if the connections happen they way you thought they should happen.